feat(authentication): document new authentication capabilities (#23)
flipt-io/docs: docs/authentication.md

---
id: "authentication"
sidebar_position: 7
---

# Authentication

Flipt supports the ability to secure its core API routes.

:::info
Flipt authentication is **disabled** (not required) by default.

Head to the [Configuration: Authentication](./configuration#authentication) section to enable it.
:::

Once enabled, all routes beneath the following API prefixes will require a [client token](#client-tokens) in order to authenticate requests:

- `/api/v1/`
- `/auth/v1/`

### Client Tokens

Client tokens are the core credential required to authenticate a request.
Tokens themselves are acquired via [authentication methods](#authentication-methods).

> Flipt will ultimately support multiple authentication methods.
> Today we only support the [token](#token) authentication method.

Reviewer: ```suggestion
Flipt will ultimately support multiple authentication methods, however only the [token](#token) authentication method is supported as of now.
```

The lower-case `authorization` metadata key should be supplied with a single string `Bearer <client-token>` to any RPC calls.

**Example (in Go):**

The following example authenticates a single gRPC client request:
```go
ctx := metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer gt6P_zIqTnCngfHDCpWb48ob5EBt3PqunUhpofNCNnc=")

flag, err := flipt.GetFlags(ctx, &flipt.GetFlagRequest{
  Key: flagKey,
})
```

This subsequent example demonstrates using a client unary interceptor, which authenticates all outgoing requests:
```go
func AuthUnaryClientInterceptor(optFuncs ...CallOption) grpc.UnaryClientInterceptor {
    return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
        ctx = metadata.AppendToOutgoingContext(ctx, "authorization", "Bearer gt6P_zIqTnCngfHDCpWb48ob5EBt3PqunUhpofNCNnc=")
        return invoker(ctx, method, req, reply, cc, opts...)
	}
}
```

### Authentication Methods

:::tip
See [Configuration: Authentication Methods](./configuration#authentication-methods) for details on how to configure the various authentication methods.
:::

#### Token

:::note

> When `token` method is enabled and zero tokens exist in the backing store, an initial token is created and logged in Flipt's server output.

Reviewer: ```suggestion
When `token` method is enabled and no tokens exist in the backing store, an initial token is created and logged in Flipt's server output.
```
